frontend/javascripts/oxalis/model/sagas/save_saga.js

```diff
@@ -216,7 +232,23 @@ function* compressBucketAndAddToUndoBatch(
   );
   const compressedBucketData = yield* call(byteArrayToLz4Array, bucketDataAsByteArray, true);
   if (compressedBucketData != null) {
-    undoBatch.push({ zoomedBucketAddress, data: compressedBucketData });
+    const volumeUndoPart: UndoBucket = {
+      zoomedBucketAddress,
+      data: compressedBucketData,
+      maybeBucketLoadedPromise,
+    };
+    if (maybeBucketLoadedPromise != null) {
+      maybeBucketLoadedPromise.then(async backendBucketData => {
+        const backendDataAsByteArray = new Uint8Array(
+          backendBucketData.buffer,
+          backendBucketData.byteOffset,
+          backendBucketData.byteLength,
+        );
+        const compressedBackendData = await byteArrayToLz4Array(backendDataAsByteArray, true);
+        volumeUndoPart.backendData = compressedBackendData;
```

Comment:
I'm struggling to wrap my head around all of this, but wouldn't it be possible to do the merging which is done in line 348 and line 367, instead here, directly? In my head, this would greatly simplify the `applyAndGetRevertingVolumeBatch`, because no special treatment would need to be implemented, there (except of maybe waiting for the promise to be resolved). But it could very well be that I'm overlooking something. Let me know if it's easier to discuss this in a call.

Reply:
> In my head, this would greatly simplify the applyAndGetRevertingVolumeBatch, because no special treatment would need to be implemented, there (except maybe waiting for the promise to be resolved). But it could very well be that I'm overlooking something. Let me know if it's easier to discuss this in a call.


Indeed, it is possible to do the merging once the backend data arrived. But we decided against this to defer the more intensive computation. If the backend data is merged directly once it arrives, the following will be done for each bucket, whose backend data is being fetched, when the user drew over it:
Once the backend data is fetched:
- decompress the current "revert"-bucket data,
- merge it with the backend data,
- compress the result again

And if the user then undos the action:
- decompress the bucket data 
- apply the decompressed data

The way the code works now is more complicated but should be (not measured) more efficient in normal uses cases. Because we assume that only a few brush strokes are undone by the user and therefore only a few buckets actually need to be merged with the backend data:
Once the backend data is fetched:
- compress the backend data

And only if the undo action is triggered to the following:
- decompress the backend data and the "normal revert"-bucket data
- merge them
- apply the merged data


As you can see, in the current implementation there are some steps saved in case that the volume action is not undone.

Does this explain the design decision to you and do you think this is justified?

Reply:
Thanks a lot for the thorough explanation! It does make sense now that I understand the reasoning :) Could you add a comment summarizing this rationale above line 248?